### PR TITLE
Support Unicode in lexer

### PR DESCRIPTION
**Pull request by sand1k**

### CODE REVIEW

File: jerry-core/parser/js/lexer.cpp [hunk 1]
```diff
@@ -1310,30 +1503,38 @@ lexer_prev_token (void)
...
+  JERRY_ASSERT (is_token_parse_in_progress == false);
 
-  buffer = buffer_start + locus;
+  lit_utf8_iterator_seek (&src_iter, locus);
   saved_token = empty_token;
 }
 
 void
-lexer_locus_to_line_and_column (size_t locus, size_t *line, size_t *column)
+lexer_locus_to_line_and_column (lit_utf8_iterator_pos_t locus, size_t *line, size_t *column)
```
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: `line_p`, `column_p`

File: jerry-core/parser/js/lexer.cpp [hunk 2]
```diff
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#include <lit/lit-strings.h>
```
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: `"lit-strings.h"`

File: jerry-core/parser/js/lexer.cpp (hunk 1 above)
Comment (egavrin) [inline jerry-core/parser/js/lexer.cpp]: Missing comments

**Comment by sand1k** [comment]
@egavrin, I've updated the PR.

**Comment by egavrin** [comment]
Good to me.

File: jerry-core/lit/lit-char-helpers.cpp [hunk 3]
```diff
@@ -253,36 +253,78 @@ lit_char_is_unicode_connector_punctuation (ecma_char_t c) /**< code unit */
...
+lit_char_is_octal_digit (ecma_char_t c) /**< code unit */
+{
+  if (c >= LIT_CHAR_ASCII_OCTAL_DIGITS_BEGIN && c <= LIT_CHAR_ASCII_OCTAL_DIGITS_END)
+  {
+    return true;
+  }
+  else
+  {
+    return false;
+  }
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.cpp]: return (c >= LIT_CHAR_ASCII_OCTAL_DIGITS_BEGIN && c <= LIT_CHAR_ASCII_OCTAL_DIGITS_END);

File: jerry-core/lit/lit-char-helpers.cpp [hunk 4]
```diff
@@ -253,36 +253,78 @@ lit_char_is_unicode_connector_punctuation (ecma_char_t c) /**< code unit */
...
+    return true;
+  }
+  else if (c >= LIT_CHAR_ASCII_UPPERCASE_LETTERS_HEX_BEGIN && c <= LIT_CHAR_ASCII_UPPERCASE_LETTERS_HEX_END)
+  {
+    return true;
+  }
+  else
+  {
+    return false;
+  }
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.cpp]: What about using || operator to connect the true cases?

File: jerry-core/lit/lit-char-helpers.cpp [hunk 5]
```diff
@@ -293,8 +335,28 @@ lit_char_hex_to_int (ecma_char_t c) /**< code unit, corresponding to
...
+{
+  if ((c >= LIT_CHAR_ASCII_LOWERCASE_LETTERS_BEGIN && c <= LIT_CHAR_ASCII_LOWERCASE_LETTERS_END)
+      || (c >= LIT_CHAR_ASCII_UPPERCASE_LETTERS_BEGIN && c <= LIT_CHAR_ASCII_UPPERCASE_LETTERS_END)
+      || (c >= LIT_CHAR_ASCII_DIGITS_BEGIN && c <= LIT_CHAR_ASCII_DIGITS_END)
+      || c == LIT_CHAR_UNDERSCORE)
+  {
+    return true;
+  }
+
+  return false;
```
Comment (zherczeg) [inline jerry-core/lit/lit-char-helpers.cpp]: The style here is different than other places (no else).

File: jerry-core/lit/lit-strings.cpp [hunk 6]
```diff
@@ -118,6 +154,28 @@ lit_is_utf8_string_valid (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string *
...
 /**
+ * Check if the code unit type is low surrogate
+ *
+ * @return true / false
+ */
+bool
+lit_is_code_unit_low_surrogate (ecma_char_t code_unit) /**< code unit */
+{
+  return LIT_UTF16_LOW_SURROGATE_MIN <= code_unit && code_unit <= LIT_UTF16_LOW_SURROGATE_MAX;
+} /* lit_is_code_unit_low_surrogate */
```
Comment (zherczeg) [inline jerry-core/lit/lit-strings.cpp]: This style is preferred by me.

File: jerry-core/lit/lit-strings.cpp [hunk 7]
```diff
@@ -757,3 +834,20 @@ bool lit_compare_utf8_strings_relational (const lit_utf8_byte_t *string1_p, /**<
...
   return (lit_utf8_iterator_is_eos (&iter1) && !lit_utf8_iterator_is_eos (&iter2));
 } /* lit_compare_utf8_strings_relational */
+
+/**
+ * Print code unit to standard output
+ */
+void
+lit_put_ecma_char (ecma_char_t ecma_char) /**< code unit */
+{
+  if (ecma_char <= 0x7f)
```
Comment (zherczeg) [inline jerry-core/lit/lit-strings.cpp]: There is a constant for this in lit_string.h.

File: jerry-core/lit/lit-char-helpers.cpp [hunk 8]
```diff
@@ -253,36 +253,78 @@ lit_char_is_unicode_connector_punctuation (ecma_char_t c) /**< code unit */
...
+/**
+ * Check if specified character is one of DecimalDigit characters (ECMA-262 v5, 7.8.3)
  *
- * @return true - if the character is a word character
- *         false - otherwise.
+ * @return true / false
  */
 bool
-lit_char_is_word_char (ecma_char_t c) /**< code unit */
+lit_char_is_decimal_digit (ecma_char_t c) /**< code unit */
```
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.cpp]: @sand1k , `Date.parse` is on master now and it's using `lit_char_is_unicode_digit`. If you introduce `lit_char_is_decimal_digit` in this patch, then could you change the `lit_char_is_unicode_digit` call in `ecma-builtin-date.cpp` to `lit_char_is_decimal_digit` in your patch too?

File: jerry-core/lit/lit-char-helpers.cpp (hunk 5 above)
Comment (LaszloLango) [inline jerry-core/lit/lit-char-helpers.cpp]: I think this is the correct style. Other places should be fixed.

File: jerry-core/parser/js/lexer.cpp [hunk 9]
```diff
@@ -1251,15 +1445,15 @@ lexer_next_token_private (void)
...
-}
+
+  PARSE_ERROR ("Illegal character", lit_utf8_iterator_get_pos (&src_iter));
+} /* lexer_parse_token */
 
 token
 lexer_next_token (void)
 {
-  if (buffer == buffer_start)
+  if (lit_utf8_iterator_get_pos (&src_iter).offset == 0)
```
Comment (zherczeg) [inline jerry-core/parser/js/lexer.cpp]: Shouldn't we need to check that we are not at a low surrogate?

**Comment by zherczeg** [comment]
Some style issues, but the patch is close to ready.

File: jerry-core/lit/lit-char-helpers.cpp (hunk 8 above)
Comment (sand1k) [inline jerry-core/lit/lit-char-helpers.cpp]: Sure.

**Comment by sand1k** [comment]
@zherczeg, the PR was updated, please, check.

**Comment by egavrin** [comment]
Even better ^_^

**Comment by zherczeg** [comment]
+1 lgtm

**Comment by egavrin** [comment]
`make push`

**Comment by egavrin** [comment]
Merged as 146ac15b5ffab8a08b45276cab9ab109ec6bd254

**Pull request merged**